programs/x509/cert_write.c

```diff
@@ -517,6 +541,68 @@ int main(int argc, char *argv[])
                 *tail = ext_key_usage;
                 tail = &ext_key_usage->next;
 
+                q = r;
+            }
+        } else if (strcmp(p, "san") == 0) {
+            prev = NULL;
+
+            while (q != NULL) {
+                if ((r = strchr(q, ';')) != NULL) {
+                    *r++ = '\0';
+                }
+
+                cur = mbedtls_calloc(1, sizeof(mbedtls_x509_san_list));
+                if (cur == NULL) {
+                    mbedtls_printf("Not enough memory for subjectAltName list\n");
+                    goto usage;
+                }
+
+                cur->next = NULL;
+
+                if ((r2 = strchr(q, ':')) != NULL) {
+                    *r2++ = '\0';
+                }
```

Comment:
URIs typically contain a colon, and I believe DirectoryName can contain just about anything.  Therefore, there needs to be a way to escape the delimiter.

Reply:
This is just an example program, not a comprehensive implementation. While I understand that it is tempting to supply it with more functionality, simplicity is also a factor. Do you think that having support for a ":" in the directoryName of a certificate generated will give us a valuable test case? So far we only have one test for a certificate with a subjectAltName generated from cert_write - see [here](https://github.com/Mbed-TLS/mbedtls/pull/7400/files#diff-c882a142e0059a06b9f052239bff1250881537adaeb385351ce03365fce69e98R164).

Reply:
> So if the documentation says you can pass a URI, there should be a way to pass a colon.

OK, I looked again, and there's really no problem here. What we're looking at here is the code searching for ":" to determine what subtype of SubjectAltName it's currently parsing, whether it's URI, DNS, DN, etc. That's just the first step, and first identifier to find.
We can have colons in any field later on and it works fine. The example in PR description already has such field, but we can go further:
`./cert_write selfsign=1 is_ca=1 issuer_key=../../tests/data_files/server8.key san="DN:C=https://github.com,O=\"Mbed TLS\",CN=\"SubjectAltName test\";URI:http://pki.example.com/;IP:127.1.1.0;DNS:example.com"`
Gives us (openssl parsing):
![image](https://user-images.githubusercontent.com/29195988/234338780-dae65a43-f3a1-408f-97a0-5815230a2e06.png)

Reply:
Semicolons are valid in URIs.  You can see them in many cgit URLs, for example.